presto-main/src/main/java/com/facebook/presto/operator/PagesSpatialIndex.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.esri.core.geometry.ogc.OGCGeometry;
+import com.facebook.presto.Session;
+import com.facebook.presto.geospatial.GeometryUtils;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.gen.JoinFilterFunctionCompiler.JoinFilterFunctionFactory;
+import com.google.common.collect.ImmutableList;
+import com.vividsolutions.jts.geom.Envelope;
+import com.vividsolutions.jts.index.strtree.AbstractNode;
+import com.vividsolutions.jts.index.strtree.ItemBoundable;
+import com.vividsolutions.jts.index.strtree.STRtree;
```

Comment:
Have we analyzed the implementation of SRTree? We need to ensure that:
1) we can get reasonable accurate estimate of its size to be able to track memory usage
2) the implementation won’t cause issues for the garbage collector (complex graph of long-lived objects and linkages, etc)

Reply:
Do you have any tips on how to do that? E.g. how can I check whether the estimated memory size is close to the actual memory size? What can I do to ensure #2? (The structure of the STRtree object looks a bit simpler than the structure of the OGCGeometry objects.)

Reply:
@martint and I discussed this "offline". Martin pointed out that STRTree implementation uses a tree of Java objects which may cause GC problems if there are a lot of nodes. At the moment, given that the largest query I'm looking at has only 15K objects with 16K nodes in the tree we think it will be ok. In the future, it would be desirable to rewrite the implementation to use flat arrays similar to how QuantileDigest was rewritten: 

- before: https://github.com/airlift/airlift/blob/0fee8406c4f013128bd0b3983738a0156ed1e3fb/stats/src/main/java/io/airlift/stats/QuantileDigest.java

- after: https://github.com/airlift/airlift/blob/master/stats/src/main/java/io/airlift/stats/QuantileDigest.java